src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs

```diff
@@ -173,14 +173,16 @@ public override void Visit(ITypeDefinition parentType, IEnumerable<IFieldDefinit
                     // taking pointers to this struct because the GC will not track updating those references
                     if (hasRefPrivateField)
                     {
-                        DummyFieldWriterHelper(parentType, excludedFields, newFields, parentType.PlatformType.SystemObject);
+                        IFieldDefinition fieldType = DummyFieldWriterHelper(parentType, excludedFields, parentType.PlatformType.SystemObject);
+                        newFields.Add(fieldType);
                     }
 
-                    bool hasValueTypePrivateField = excludedFields.Any(f => f.Type.IsValueType);
+                    bool hasValueTypePrivateField = excludedFields.Any(f => !f.Type.IsOrContainsReferenceType());
```

Comment:
I'm not sure how expensive this is but I wonder if instead you just filter the list while checking for the private ref fields and this will just be what remains.

Reply:
I don't know how that would by faster while still using linq expressions and IEnumerable. We still have to first check if we have any private ref fields. So what exactly would we be filtering during the first pass?

I could do something like this, but I doubt it would be much faster:
```C#
bool hasRefPrivateField = excludedFields.Any(f => f.Type.IsOrContainsReferenceType());

var fieldsLeft = excludedFields.Where(f => !f.Type.IsOrContainsReferenceType());
if (fieldsLeft.Count() > 0)
{
    // ...
}
```

Reply:
How about something like this:
```C#
int count = 0;
foreach (IFieldDefinition field in excludedFields)
{
    if (field.Type.IsOrContainsReferenceType())
    {
        hasRefPrivateField = true;
        break;
    }
    count++;
}

var leftoverFields = excludedFields.Skip(count);
if (count > 0)
{
    // ...
}
```

Reply:
Don't worry about it for now. I don't think it is going to matter as the number of fields is pretty small.